File: backend/app/schemas/factor.py
"""因子分析 schemas。"""
from __future__ import annotations

from pydantic import BaseModel, Field


class FactorICRow(BaseModel):
    field: str
    label: str
    ic_return: float | None = None
    ic_max_up: float | None = None


class QuantileRow(BaseModel):
    quantile: str
    count: int
    mean: float
    median: float
    win_rate: float
    big_win_rate: float


class FactorQuantile(BaseModel):
    field: str
    label: str
    quantiles: list[QuantileRow]


class FactorAnalysisResponse(BaseModel):
    task_id: str
    total_trades: int
    ic: list[FactorICRow]
    quantiles: list[FactorQuantile]


class FactorAnalysisRequest(BaseModel):
    task_id: str
    target: str = Field(default="return_pct", description="目标变量：return_pct 或 max_up_pct")
    quantile_n: int = Field(default=5, ge=2, le=10)

File: Web/src/api/factor.ts
import { api } from "./client";

export interface FactorICRow {
  field: string;
  label: string;
  ic_return: number | null;
  ic_max_up: number | null;
}

export interface QuantileRow {
  quantile: string;
  count: number;
  mean: number;
  median: number;
  win_rate: number;
  big_win_rate: number;
}

export interface FactorQuantile {
  field: string;
  label: string;
  quantiles: QuantileRow[];
}

export interface FactorAnalysisResponse {
  task_id: string;
  total_trades: number;
  ic: FactorICRow[];
  quantiles: FactorQuantile[];
}

export async function getFactorAnalysis(
  taskId: string,
  opts: { target?: "return_pct" | "max_up_pct"; quantileN?: number } = {}
): Promise<FactorAnalysisResponse> {
  const { data } = await api.get<FactorAnalysisResponse>("/factor/analysis", {
    params: {
      task_id: taskId,
      target: opts.target ?? "return_pct",
      quantile_n: opts.quantileN ?? 5,
    },
  });
  return data;
}
